server/routes/api/import_detection/detection.py

```diff
@@ -48,6 +57,63 @@ def _detect_date(col_order: List[Column],
     return None
 
 
+def _detect_place(col_order: List[Column],
+                  cols_sampled: Dict[int, List[str]]) -> Optional[MappingVal]:
+    """Process all columns and return the one which best represents the
+    detected Place along with its details. If no Place is detected, the
+    return value is None.
+    @args:
+        col_order: A list of Column objects where the list order signifies the column indices.
+        cols_sampled: A mapping from column index to a list of strings.
+    @returns:
+        The detected MappingVal structure or None if no Place is detected.
+    """
+    # Currently, only country and state detection is supported.
+    # detected_states and detected_countries are mappings from the detected
+    # column's index to the TypeProperty of the detected Place.
+    detected_states: Dict[int, TypeProperty] = {}
+    detected_countries: Dict[int, TypeProperty] = {}
+
+    for col in col_order:
+        col_header = col.header
+        col_ind = col.column_index
+        vals = cols_sampled[col_ind]
+
+        place: Optional[
+            TypeProperty] = place_detector.detect_column_with_places(
+                col_header, vals)
+        if place is None:
+            continue
+
+        if place.dc_type.dcid == "Country":
+            detected_countries[col_ind] = place
+        elif place.dc_type.dcid == "State":
+            detected_states[col_ind] = place
+
+    # Now get the preferred property detected.
+    # Note that State detection is given preference over Country detection
+    # because it is more precise.
+    detected_places: Dict[int, TypeProperty] = {}
+    prop_order: List[str] = {}
+    if detected_states:
+        detected_places = detected_states
+        prop_order = place_detector.STATE_PROP_PREF_ORDER
+    elif detected_countries:
+        detected_places = detected_countries
+        prop_order = place_detector.COUNTRY_PROP_PREF_ORDER
+
+    preferred_ind: Optional[int] = place_detector.preferred_property(
```

Comment:
nit: preferred_idx - took me a bit to figure out what "ind" was supposed to stand for

Reply:
Updated here and elsewhere.